format-config-v1.go

```diff
@@ -37,66 +43,191 @@ type formatConfigV1 struct {
 	XL      *xlFormat `json:"xl,omitempty"`
 }
 
-// FIXME: currently we don't check single exportPath which uses FS layer.
+// validateXLJBOD - validate xl jbod order if they are consistent.
+func validateXLJBOD(jbod []string, jsonMap map[int]formatConfigV1) error {
+	jbodStr := strings.Join(jbod, ", ")
+	for _, format := range jsonMap {
+		if format.XL == nil {
+			continue
+		}
+		savedJBODStr := strings.Join(format.XL.JBOD, ", ")
+		if jbodStr != savedJBODStr {
+			return errors.New("Inconsistent disks.")
+		}
+	}
+	return nil
+}
 
-// loadFormatXL - load XL format.json.
-func loadFormatXL(storage StorageAPI) (xl *xlFormat, err error) {
-	offset := int64(0)
-	r, err := storage.ReadFile(minioMetaBucket, formatConfigFile, offset)
-	if err != nil {
-		return nil, err
+// arrangeDiskInOrder - arrange disks in saved order.
+func arrangeDiskInOrder(storageDisks []StorageAPI, jbodMap map[int]formatConfigV1) ([]StorageAPI, error) {
+	var savedJBOD []string
+	for index := range storageDisks {
+		if format, ok := jbodMap[index]; ok {
+			if format.XL == nil {
+				continue
+			}
+			savedJBOD = format.XL.JBOD
+			break
+		}
 	}
-	decoder := json.NewDecoder(r)
-	formatXL := formatConfigV1{}
-	err = decoder.Decode(&formatXL)
-	if err != nil {
+	if err := validateXLJBOD(savedJBOD, jbodMap); err != nil {
 		return nil, err
 	}
-	if err = r.Close(); err != nil {
-		return nil, err
+
+	// Pick the first JBOD list to verify the order and construct new
+	// set of disk slice.
+	var newDisks = make([]StorageAPI, len(storageDisks))
+	for dIndex := range storageDisks {
+		for index, uuid := range savedJBOD {
+			if jbodMap[dIndex].XL == nil {
+				continue
+			}
+			if jbodMap[dIndex].XL.Disk == uuid {
+				newDisks[index] = storageDisks[index]
+			}
+		}
 	}
-	if formatXL.Version != "1" {
-		return nil, fmt.Errorf("Unsupported version of backend format [%s] found.", formatXL.Version)
+	for index := range newDisks {
+		if newDisks[index] == nil {
+			newDisks[index] = storageDisks[index]
+		}
 	}
-	if formatXL.Format != "xl" {
-		return nil, fmt.Errorf("Unsupported backend format [%s] found.", formatXL.Format)
+	return newDisks, nil
+}
+
+// loadFormatXL - load XL format.json.
+func loadFormatXL(storageDisks ...StorageAPI) (jbodMap map[int]formatConfigV1, err error) {
+	var fileNotFoundCount = 0
```

Comment:
> func loadFormatXL(storageDisks ...StorageAPI) (formatConfigs map[string]formatConfigV1, err error) {
> 
> where index is disk

disk is a string, i avoided that by not passing in the disks argument. So kept the index which is also correct. Just indicates the index of first disk, second disk ..